core/jobs/batch_jobs/exp_model_validation_jobs.py

```diff
@@ -0,0 +1,134 @@
+
+# coding: utf-8
+#
+# Copyright 2022 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Validation Jobs for exploration model"""
+
+from __future__ import annotations
+
+from core.jobs import base_jobs
+from core.jobs.io import ndb_io
+from core.jobs.transforms import job_result_transforms
+from core.jobs.types import job_run_result
+from core.platform import models
+
+import apache_beam as beam
+import bs4
+from typing import Iterator
+
+(exp_models, ) = models.Registry.import_models([models.NAMES.exploration])
+
+
+# https://stackoverflow.com/questions/39233973/get-all-keys-of-a-nested-dictionary  # pylint: disable=line-too-long
+def recursive_items(dictionary: dict) -> Iterator[tuple]:
+    """Yields an iterator containing tuples of key, value pairs
+
+        Yields:
+            tuple. Yields tuples of key, value pairs.
+        """
+    for key, value in dictionary.items():
+        if isinstance(value, dict):
+            yield from recursive_items(value)
+        else:
+            yield (key, value)
+
+
+def get_invalid_links(dictionary: dict) -> list:
+    """Returns list of invalid links
+
+        Returns:
+            list. Returns list of invalid links.
+        """
+    string = ''
+    for key, value in recursive_items(dictionary):
+        if key == 'html':
+            string += value
+
+    soup = bs4.BeautifulSoup(string, 'html.parser')
+
+    # Extract links.
+    links = soup.find_all('oppia-noninteractive-link')
+    cleaned_links = []
+    for link in links:
+        # Remove &quot; from the links.
+        cleaned_links.append(link.get('url-with-value')[6:-6])
+
+    invalid_links = []
+    # Extract scheme.
+    for link in cleaned_links:
+        link_info = link.split(':')
```

Comment:
Why is this done?

Reply:
This is done to separate the scheme from the link URL. For example, if the link is 'https://www.example.com", we are only interested in checking the scheme i.e. 'https' in this case. If the scheme is not available (www.test.com), the default scheme (https) is used.

Reply:
From your comment it is a bit hard to read, but I get what you mean, can you explain it in a code comment.

Reply:
Done! I've added the comments.